### PR TITLE
Close #12098: Use ScreenRect on gfx_filter_rect

### LINKED ISSUE #12098: Use ScreenRect on gfx_filter_rect
- Create an overload of `gfx_filter_rect` with the same name, but that takes a `const ScreenRect&` object instead of `left`, `top`, `right`, `bottom`
- Make the old function call the new one to avoid code duplication
- Replace the calls to the one using the 4 variables by the one using `ScreenRect`.
- If no calls to the old one are left, remove it.

*Note*: Don't try to update all calls at once, unless it doesn't occur a lot, as this might lead to a huge PR.

### PR DESCRIPTION
**Pull request by annoyingdroid**
Refactored `gfx_filter_rect` to use `const ScreenRect&` and modified calls to the old function to make them point at the new one.

### CODE REVIEW

File: src/openrct2-ui/windows/Tooltip.cpp [hunk 1]
```diff
@@ -193,14 +193,14 @@ static void window_tooltip_paint(rct_window* w, rct_drawpixelinfo* dpi)
     int32_t bottom = w->windowPos.y + w->height - 1;
 
     // Background
-    gfx_filter_rect(dpi, left + 1, top + 1, right - 1, bottom - 1, PALETTE_45);
-    gfx_filter_rect(dpi, left + 1, top + 1, right - 1, bottom - 1, PALETTE_GLASS_LIGHT_ORANGE);
+    gfx_filter_rect(dpi, { left + 1, top + 1, right - 1, bottom - 1 }, PALETTE_45);
+    gfx_filter_rect(dpi, { left + 1, top + 1, right - 1, bottom - 1 }, PALETTE_GLASS_LIGHT_ORANGE);
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Tooltip.cpp:197]:
```suggestion
    gfx_filter_rect(dpi, { { left + 1, top + 1 }, { right - 1, bottom - 1 } }, PALETTE_45);
    gfx_filter_rect(dpi, { { left + 1, top + 1 }, { right - 1, bottom - 1 } }, PALETTE_GLASS_LIGHT_ORANGE);
```

File: src/openrct2-ui/windows/Tooltip.cpp [hunk 2]
```diff
@@ -193,14 +193,14 @@ static void window_tooltip_paint(rct_window* w, rct_drawpixelinfo* dpi)
...
 
     // Sides
-    gfx_filter_rect(dpi, left + 0, top + 2, left + 0, bottom - 2, PALETTE_DARKEN_3);
-    gfx_filter_rect(dpi, right + 0, top + 2, right + 0, bottom - 2, PALETTE_DARKEN_3);
-    gfx_filter_rect(dpi, left + 2, bottom + 0, right - 2, bottom + 0, PALETTE_DARKEN_3);
-    gfx_filter_rect(dpi, left + 2, top + 0, right - 2, top + 0, PALETTE_DARKEN_3);
+    gfx_filter_rect(dpi, { left + 0, top + 2, left + 0, bottom - 2 }, PALETTE_DARKEN_3);
+    gfx_filter_rect(dpi, { right + 0, top + 2, right + 0, bottom - 2 }, PALETTE_DARKEN_3);
+    gfx_filter_rect(dpi, { left + 2, bottom + 0, right - 2, bottom + 0 }, PALETTE_DARKEN_3);
+    gfx_filter_rect(dpi, { left + 2, top + 0, right - 2, top + 0 }, PALETTE_DARKEN_3);
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Tooltip.cpp:203]:
```suggestion
    gfx_filter_rect(dpi, { { left + 0, top + 2 }, { left + 0, bottom - 2 } }, PALETTE_DARKEN_3);
    gfx_filter_rect(dpi, { { right + 0, top + 2 }, { right + 0, bottom - 2 } }, PALETTE_DARKEN_3);
    gfx_filter_rect(dpi, { { left + 2, bottom + 0 }, { right - 2, bottom + 0 } }, PALETTE_DARKEN_3);
    gfx_filter_rect(dpi, { { left + 2, top + 0 }, { right - 2, top + 0 } }, PALETTE_DARKEN_3);
```

File: src/openrct2/drawing/Drawing.cpp [hunk 3]
```diff
@@ -585,7 +585,7 @@ void gfx_draw_pixel(rct_drawpixelinfo* dpi, const ScreenCoordsXY& coords, int32_
 
 void gfx_filter_pixel(rct_drawpixelinfo* dpi, const ScreenCoordsXY& coords, FILTER_PALETTE_ID palette)
 {
-    gfx_filter_rect(dpi, coords.x, coords.y, coords.x, coords.y, palette);
+    gfx_filter_rect(dpi, { coords.x, coords.y, coords.x, coords.y }, palette);
```
Comment (tupaschoal) [inline src/openrct2/drawing/Drawing.cpp:588]:
```suggestion
    gfx_filter_rect(dpi, { coords, coords }, palette);
```

File: src/openrct2/interface/Chat.cpp [hunk 4]
```diff
@@ -131,7 +131,7 @@ void chat_draw(rct_drawpixelinfo* dpi, uint8_t chatBackgroundColor)
         }
 
         gfx_set_dirty_blocks(_chatLeft, _chatTop - 5, _chatRight, _chatBottom + 5);             // Background area + Textbox
-        gfx_filter_rect(dpi, _chatLeft, _chatTop - 5, _chatRight, _chatBottom + 5, PALETTE_51); // Opaque gray background
+        gfx_filter_rect(dpi, { _chatLeft, _chatTop - 5, _chatRight, _chatBottom + 5 }, PALETTE_51); // Opaque gray background
```
Comment (tupaschoal) [inline src/openrct2/interface/Chat.cpp:134]:
```suggestion
        gfx_filter_rect(dpi, { { _chatLeft, _chatTop - 5 }, { _chatRight, _chatBottom + 5 } }, PALETTE_51); // Opaque gray background
```

**Comment by tupaschoal** [comment]
Make sure to look at the code formatting issue too and fix it :)

**Comment by annoyingdroid** [comment]
> Does this mean that there are no callers to the original function? If so, you can delete it.

There are still quite a few calls to the original function (more than I'd feel comfortable changing all at once) so I've only done a handful for the time. The old function does reference the new one though. 

I'll work on getting the code formatting fixed!

**Comment by tupaschoal** [comment]
> There are still quite a few calls to the original function (more than I'd feel comfortable changing all at once) so I've only done a handful for the time. The old function does reference the new one though.

Then you took the right approach, thanks!

Note that there is a merge conflict due to recently merged code, you'll have to rebase your PR. To do so:
- `git checkout develop`
- `git pull upstream develop --rebase`
- `git checkout gfx_filter_rect`
- `git rebase -i develop`
- Follow the instructions to solve the merge conflicts
- `git push -f origin`

**Comment by annoyingdroid** [comment]
Do you have any tips on addressing the code formatting? I read the formatting guidelines and thought I had sorted the issue but the check still fails.

**Comment by tupaschoal** [comment]
I renamed it as "Part of #..." instead of "Close #..." as you mentioned that there are quite a few calls yet and the old function still lives.

**Pull request merged**